airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/StateFilesAccessor.kt

```diff
@@ -0,0 +1,192 @@
+import org.apache.kafka.connect.json.JsonConverterConfig
+import org.apache.kafka.connect.runtime.WorkerConfig
+import org.apache.kafka.connect.runtime.standalone.StandaloneConfig
+import org.apache.kafka.connect.storage.Converter
+import org.apache.kafka.connect.storage.FileOffsetBackingStore
+
+/**
+ * [StateFilesAccessor] reads and writes from Debezium offset and schema history files. This should
+ * NOT be done while the Debezium Engine is running! None of this is thread-safe. Instead, this
+ * object should be used to:
+ * 1. create and populate the files prior to running the Debezium Engine,
+ * 2. read the changed files after running the Debezium Engine.
+ *
+ * The files are deleted on [close].
+ */
+class StateFilesAccessor : AutoCloseable {
+
+    val offsetFilePath: Path
+    val schemaFilePath: Path
+
+    internal val workingDir: Path
+    private val fileOffsetBackingStore: FileOffsetBackingStore
+    private val fileSchemaHistory: FileSchemaHistory
+
+    override fun close() {
+        fileOffsetBackingStore.stop()
+        fileSchemaHistory.stop()
+        try {
+            FileUtils.deleteDirectory(workingDir.toFile())
+        } catch (e: IOException) {
+            throw UncheckedIOException(e)
+        }
+    }
+
+    fun readUpdatedOffset(offset: DebeziumComponent.State.Offset): DebeziumComponent.State.Offset {
+        val bytesKeys = offset.debeziumOffset.keys.map { key -> toByteBuffer(key) }
+        fileOffsetBackingStore.start()
+        val future = fileOffsetBackingStore.get(bytesKeys)
+        fileOffsetBackingStore.stop()
+        val debeziumOffset =
+            future.resultNow().map { (key, value) -> Pair(toJson(key), toJson(value)) }.toMap()
+        return DebeziumComponent.State.Offset(debeziumOffset)
+    }
+
+    fun writeOffset(offset: DebeziumComponent.State.Offset) {
+        val bytesMap =
+            offset.debeziumOffset
+                .map { (key, value) -> Pair(toByteBuffer(key), toByteBuffer(value)) }
+                .toMap()
+        fileOffsetBackingStore.start()
+        fileOffsetBackingStore.set(bytesMap, null)
+        fileOffsetBackingStore.stop()
+    }
+
+    fun readSchema(): DebeziumComponent.State.Schema {
+        val builder = ImmutableList.builder<HistoryRecord>()
+        val consumer = Consumer { element: HistoryRecord -> builder.add(element) }
+        fileSchemaHistory.start()
+        try {
+            RECOVER_RECORDS_METHOD.invoke(fileSchemaHistory, consumer)
+        } catch (e: IllegalAccessException) {
+            throw RuntimeException(e)
+        } catch (e: InvocationTargetException) {
+            throw RuntimeException(e)
+        }
+        fileSchemaHistory.stop()
+        return DebeziumComponent.State.Schema(builder.build())
+    }
+
+    fun writeSchema(schema: DebeziumComponent.State.Schema) {
+        fileSchemaHistory.initializeStorage()
+        fileSchemaHistory.start()
+        for (r in schema.debeziumSchemaHistory) {
```

Comment:
We have cases where this isn't straightforward and need to wait in some cases  to schema history to show up.
Will this be used in a way that can override for specific DB?

Reply:
I think I know what you're referring to. I haven't ported this logic yet in this stack of PRs, nor in fact have I given much thought to what shape it should take. It's a follow-up item of mine if we go forward with this.